resources/manager/data.go

```diff
@@ -62,77 +65,95 @@ func (d *Data) Output() <-chan ResourceMap {
 
 // Run updates the cache using Fetcher implementations.
 func (d *Data) Run(ctx context.Context) error {
-	updates := make(chan update)
-
-	var wg sync.WaitGroup
-	d.wg = &wg
-
-	for _, key := range d.fetchers.Keys() {
-		wg.Add(1)
-		go func(k string) {
-			defer wg.Done()
-			d.fetchWorker(ctx, updates, k)
-		}(key)
-	}
-
-	wg.Add(1)
-	go func() {
-		defer wg.Done()
-		d.fetchManager(ctx, updates)
-	}()
-
+	go d.fetchAndSleep(ctx)
 	return nil
 }
 
-func (d *Data) fetchWorker(ctx context.Context, updates chan update, k string) {
+func (d *Data) fetchAndSleep(ctx context.Context) {
+	// Happens once in a lifetime of cloudbeat and then enters the loop
+	d.fetchIteration(ctx)
 	for {
 		select {
+		case <-d.stop:
+			logp.L().Errorf("fetchers manager stopped")
+			return
 		case <-ctx.Done():
+			logp.L().Errorf("fetchers manager canceled")
 			return
-		default:
-			if !d.fetchers.ShouldRun(k) {
-				break
-			}
+		case <-time.After(d.interval):
+			d.fetchIteration(ctx)
+		}
+	}
+}
+
+func (d *Data) fetchIteration(ctx context.Context) {
+	logp.L().Infof("manager trigger fetching using %d fetchers", len(d.fetchers.Keys()))
+	d.wg = &sync.WaitGroup{}
+	mu := sync.Mutex{}
+	cycleData := make(ResourceMap)
+	start := time.Now()
 
-			val, err := d.fetchers.Run(ctx, k)
+	for _, key := range d.fetchers.Keys() {
+		d.wg.Add(1)
+		go func(k string) {
+			defer d.wg.Done()
+			val, err := d.fetchSingle(ctx, k)
 			if err != nil {
-				logp.L().Errorf("error running fetcher for key %q: %v", k, err)
+				logp.L().Errorf("error running fetcher for key %s: %v", k, err)
+			} else {
+				logp.L().Debugf("fetcher %s finished and found %d values", k, len(val))
+				mu.Lock()
+				defer mu.Unlock()
+				cycleData[k] = val
 			}
-
-			updates <- update{k, val}
-		}
-		// Go to sleep in each iteration.
-		time.Sleep(d.interval)
+		}(key)
 	}
+
+	d.wg.Wait()
+	logp.L().Infof("manager finished waiting and sending data after %d msecs", time.Since(start).Milliseconds())
+	d.output <- cycleData
```

Comment:
There are some tickets already about the processing flow (maybe we should process resources in an async matter as well).
I don't think it will be smart to get into it on this pr, as long there are no degradations as part of my pr.
WDYT?

Reply:
I'm with you